modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotFutureTask.java

```diff
@@ -0,0 +1,884 @@
+         * @param factory Factory to produce an IO interface over underlying file.
+         * @param allocPages Total number of tracking pages.
+         */
+        public PageStoreSerialWriter(
+            IgniteLogger log,
+            PageStore store,
+            BooleanSupplier checkpointComplete,
+            BooleanSupplier interrupt,
+            Consumer<Throwable> exCons,
+            File deltaFile,
+            FileIOFactory factory,
+            int allocPages,
+            ThreadLocal<ByteBuffer> localBuff
+        ) {
+            assert store != null;
+
+            this.factory = factory;
+            this.deltaFile = deltaFile;
+            this.checkpointComplete = checkpointComplete;
+            this.interrupt = interrupt;
+            this.exCons = exCons;
+            this.log = log.getLogger(PageStoreSerialWriter.class);
+            this.store = store;
+            this.localBuff = localBuff;
+            // It is important to init {@link AtomicBitSet} under the checkpoint write-lock.
+            // This guarantee us that no pages will be modified and it's safe to init pages
+            // list which needs to be processed.
+            pagesWrittenBits = new AtomicBitSet(allocPages);
+
+            store.addWriteListener(this);
+        }
+
+        /**
+         * @return {@code true} if writer is stopped and cannot write pages.
+         */
+        public boolean stopped() {
+            return (checkpointComplete.getAsBoolean() && partProcessed) || interrupt.getAsBoolean();
+        }
+
+        /**
+         * Mark partition has been processed by another thread.
+         */
+        public void markPartitionProcessed() {
+            lock.writeLock().lock();
+
+            try {
+                partProcessed = true;
+            }
+            finally {
+                lock.writeLock().unlock();
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override public void accept(long pageId, ByteBuffer buf) {
+            assert buf.position() == 0 : buf.position();
+            assert buf.order() == ByteOrder.nativeOrder() : buf.order();
+
+            if (fileIo == null) {
+                lock.writeLock().lock();
+
+                try {
+                    if (stopped())
+                        return;
+
+                    if (fileIo == null)
+                        fileIo = factory.create(deltaFile);
+                }
+                catch (IOException e) {
+                    exCons.accept(e);
+                }
+                finally {
+                    lock.writeLock().unlock();
+                }
+            }
+
+            lock.readLock().lock();
+
+            try {
+                if (stopped())
+                    return;
+
+                if (checkpointComplete.getAsBoolean()) {
+                    int pageIdx = PageIdUtils.pageIndex(pageId);
+
+                    // Page already written.
+                    if (!pagesWrittenBits.touch(pageIdx))
```

Comment:
I think we also need to check the counter of sent (by `sendPart`) pages here, otherwise, we will read and write to delta file already sent pages

Reply:
I think you are not right here. This counter and such synchronization will overcomplicate the whole process itself.
Currnetly it's a bit overhead for storing delta pages which were already sent, but no sync between sending partition file and saving delta pages.

Reply:
I'll create an issue to implement this improvement.